include/pmacc/particles/algorithm/ForEach.hpp

```diff
@@ -22,157 +22,447 @@
+             *            the rule tat all except the last frame are fully filled and in the last frame the low indices
+             *            within the frames will be contigious filled too.
              *
-             * @param areaMapperFactory factory to construct an area mapper,
-             *                          the area is defined by the constructed mapper object
+             * @tparam T_Acc alpaka accelerator type
+             * @tparam T_Functor unary particle functor or frame functor following the interface concept
+             *                           detail::FrameFunctorInterface or detail::ParticleFunctorInterface
+             * @param acc alpaka accelerator
+             * @param unaryFunctor Functor executed for each particle/frame.
+             *                     The caller must ensure that calling the functor in parallel with
+             *                     different workers is data race free.
+             *                     particle functor: 'void operator()(T_Acc const &, ParticleType)'.
+             *                       It is not allowed to call a synchronization function within the functor.
+             *                     frame functor: 'void operator()(T_Acc const &, FrameCtx)'.
+             *                       Calling synchronization within the functor is allowed.
              */
-            template<typename T_Species, typename T_Functor, typename T_AreaMapperFactory>
-            HINLINE void forEach(T_Species&& species, T_Functor functor, T_AreaMapperFactory const& areaMapperFactory)
+            template<typename T_Acc, typename T_Functor>
+            DINLINE void operator()(T_Acc const& acc, T_Functor&& unaryFunctor) const
+            {
+                if constexpr(numWorkers > frameSize)
+                {
+                    PMACC_CASSERT_MSG(
+                        __number_of_workers_must_be_a_multiple_of_framesize,
+                        (numWorkers % frameSize) == 0);
+                }
+
+                auto const& superCell = m_particlesBox.getSuperCell(m_superCellIdx);
+                uint32_t const numPartcilesInSupercell = superCell.getNumParticles();
+
+                // Information used to know if a particle within a frame is active without checking the multiMask.
+                auto frameConditionData = T_Order::createCtx(forEachParticleInFrame);
+
+                // end kernel if we have no particles
+                if(numPartcilesInSupercell == 0)
+                    return;
+
+                // get starting frame
+                auto framePtrCtx = forEachParticleInFrame(
+                    [&](lockstep::Idx const idx)
+                    {
+                        auto frame = T_Order::getFirst(m_particlesBox, m_superCellIdx);
+
+                        if constexpr(std::is_same_v<Reverse, T_Order>)
+                        {
+                            if(frame.isValid())
```

Comment:
Why is this condition needed? If we had no particles we would have returned already

Reply:
@psychocoderHPC what is your opinion on this?

Reply:
Yes I forgot about it. It is not necessarily required.
If there are no must-have changes I will fix it in a separate PR.